Adjust storage and processing in SystemPacketBuffer and associated code for large payloads.

pidarped opened the pull request:
Extend uint16_t type variables to uint32_t wherever applicable.

Fixes #31115

tcarmelveilleux commented on src/ble/BtpEngine.cpp:
```diff
@@ -310,7 +310,7 @@ CHIP_ERROR BtpEngine::HandleCharacteristicReceived(System::PacketBufferHandle &&
         // we're not truncating to a data length smaller than what we have already consumed.
         VerifyOrExit(reader.OctetsRead() <= mRxFragmentSize, err = BLE_ERROR_REASSEMBLER_INCORRECT_STATE);
-        data->SetDataLength(chip::min(data->DataLength(), mRxFragmentSize));
+        data->SetDataLength(chip::min(data->DataLength(), static_cast<uint32_t>(mRxFragmentSize)));
```
Please, everywhere, use size_t rather than uint32_t, if this is going to be a breaking change anyway?

pidarped replied:
@tcarmelveilleux, I actually have considered size_t. The issue is that there are a lot of places in both APIs and storage that need modification. Using size_t potentially requires even more instances of static_casts in the code. Furthermore, the fact that sizeof(size_t) is different in different architectures(8 bytes in 64 bit LInux, e.g.) makes handling it difficult when comparing with different other variables(ptrdiff, ssize_t, etc). I am also hitting a bunch of unit test failures. So, my theory was that since it is based off the 4 byte max field in the TCP payload, using a uint32_t should be fine. 
What is the compelling argument in favor of size_t in this particular case when we know that the max size is lower than 4 bytes?

bzbarsky-apple replied:
> What is the compelling argument in favor of size_t 

The fact that all other buffer size things, spans, etc, use it.

Realistically, you can _store_ uint32_t in packetbuffer, _return_ size_t in the API.  API consumers that for some reason don't want to use size_t can check-and-cast, but for the most part things using buffer sizes should in fact be using size_t...

pidarped replied:
>The fact that all other buffer size things, spans, etc, use it.

I believe they use it because the size of the buffers can be relatively open-ended and size_t allows them to pass in as big a size as the memory would allow it. That is not necessarily true in this case where the cap is set by the TCP payload size of 4 bytes. 
TBH, I am all for using size_t but am hitting a few issues in the Tests and also having to insert more static_casts in the code. Going to take a look again with size_t and dig into the issues.

pidarped replied:
Changed all API parameters and returns of SystemPacketBuffer to size_t. Used static_cast at higher layers wherever applicable.

tcarmelveilleux reviewed (changes requested):
Move to `size_t` rather than `uint32_t`, since this is already a disruptive change.

pidarped committed:
Adjust storage and processing in SystemPacketBuffer
 and associated code for large payloads.

Extend uint16_t type variables to size_t for the APIs
and all applicable places.

tcarmelveilleux commented on src/inet/TCPEndPointImplLwIP.cpp:
```diff
@@ -512,7 +512,7 @@ CHIP_ERROR TCPEndPointImplLwIP::AckReceive(uint16_t len)
 
     if (mTCP != nullptr)
-        tcp_recved(mTCP, len);
+        tcp_recved(mTCP, static_cast<uint16_t>(len));
```
Why a uint16_t here?

pidarped replied:
Yes, all LwIP APIs are limited to u16 which makes it problematic to deal with. So, for LwIP based systems, one solution would be to only accept TCP payloads that are less than UINT16_MAX, else return an appropriate error.

tcarmelveilleux commented on src/inet/TCPEndPointImplLwIP.cpp:
```diff
@@ -570,7 +570,7 @@ TCPEndPointImplLwIP::BufferOffset TCPEndPointImplLwIP::FindStartOfUnsent()
     {
         VerifyOrDie(!startOfUnsent.buffer.IsNull());
-        uint16_t bufDataLen = startOfUnsent.buffer->DataLength();
+        uint16_t bufDataLen = static_cast<uint16_t>(startOfUnsent.buffer->DataLength());
```
This is concerning me, in that a larger than UINT16_MAX `DataLength` would cause issues

pidarped replied:
I agree with the concern. Stumbled on the LwIP uint16 limitation midway. So, what I was thinking is to have a hard check up front on the TCP payload to limit it to less than UINT16_MAX for LwIP based systems so that all subsequent casts are de-risked. Otherwise, it would need a much more extensive change I think.

pidarped replied:
PacketBuffer::New() has a check against UINT16_MAX for LwIP. This should account for buffers created for the outgoing path. For incoming, LwIP itself would only pass a PacketBuffer that is limited to UINT16_MAX. So, the addition of the MaxTCPPayload in the CASE session params would help in informing the peer of the local system limits. https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8924

tcarmelveilleux commented on src/system/SystemPacketBuffer.h:
```diff
@@ -206,29 +206,29 @@ class DLL_EXPORT PacketBuffer : private pbuf
      */
-    void SetDataLength(uint16_t aNewLen, const PacketBufferHandle & aChainHead);
-    void SetDataLength(uint16_t aNewLen) { SetDataLength(aNewLen, nullptr); }
+    void SetDataLength(size_t aNewLen, const PacketBufferHandle & aChainHead);
+    void SetDataLength(size_t aNewLen) { SetDataLength(aNewLen, nullptr); }
```
For lwIP, looks like we can't take in > 16 bits. There needs to be range checks.

pidarped replied:
SetDataLength() implementation now has a max check for aNewLen against UNIT16_MAX for LwIP.

pidarped commented on src/system/tests/TestSystemPacketBuffer.cpp:
```diff
@@ -329,8 +329,13 @@ void PacketBufferTest::PrepareTestBuffer(BufferConfiguration * config, int flags
     }
     config->handle->payload = config->payload_ptr;
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    config->handle->len     = static_cast<uint16_t>(config->init_len);
+    config->handle->tot_len = static_cast<uint16_t>(config->init_len);
```
Added a pre check for UINT16_MAX.

tcarmelveilleux commented on src/ble/BtpEngine.cpp:
```diff
@@ -374,11 +374,11 @@ CHIP_ERROR BtpEngine::HandleCharacteristicReceived(System::PacketBufferHandle &&
     {
         // Trim remainder, if any, of the received packet buffer based on sender-specified length of reassembled message.
-        int padding = mRxBuf->DataLength() - mRxLength;
+        int padding = static_cast<uint16_t>(mRxBuf->DataLength()) - mRxLength;
```
This only works if both mRxLength and mRxBuf->DataLength() don't change values. Since this all relies on uint16_t, please pre-check "fits in uint16" at the start of this method.

The static cast here will be immediately undone due to customary integer promotion runs.

yufengwangca replied:
+1

pidarped replied:
For BLE, DataLength() is limited to UINT16_MAX since the length field of payload is 2 bytes long instead of 4 bytes(TCP). For processing BLE packets, in this case, there should not be any data loss.

tcarmelveilleux commented on src/ble/BtpEngine.cpp:
```diff
@@ -403,7 +403,7 @@ CHIP_ERROR BtpEngine::HandleCharacteristicReceived(System::PacketBufferHandle &&
         if (!mRxBuf.IsNull())
         {
-            ChipLogError(Ble, "With rx buf data length = %u", mRxBuf->DataLength());
+            ChipLogError(Ble, "With rx buf data length = %u", static_cast<uint16_t>(mRxBuf->DataLength()));
```
%u is unsigned. Cast to unsigned.

yufengwangca replied:
Instead of casting to uint16_t for printing,  use the format specifiers defined in <inttypes.h> for right fixed-width integers. See example above.

pidarped replied:
Can go either way, i.e., unsigned or PRIu16.

tcarmelveilleux commented on src/ble/BtpEngine.cpp:
```diff
@@ -456,7 +456,7 @@ bool BtpEngine::HandleCharacteristicSend(System::PacketBufferHandle data, bool s
         mTxBuf    = std::move(data);
         mTxState  = kState_InProgress;
-        mTxLength = mTxBuf->DataLength();
+        mTxLength = static_cast<uint16_t>(mTxBuf->DataLength());
```
Must ensure no overrun.

pidarped replied:
For BLE packets, there should be no data loss with a cast to uint16_t.

tcarmelveilleux commented on src/inet/tests/TestInetLayerCommon.cpp:
```diff
@@ -135,18 +135,18 @@ static void FillDataBufferPattern(uint8_t * aBuffer, uint16_t aLength, uint16_t
                    "expected 0x%02x, found 0x%02x\n",
-                   i, i, aFirstValue, lValue);
-            DumpMemory(aBuffer + aPatternStartOffset, aLength - aPatternStartOffset, "0x", 16);
+                   static_cast<uint16_t>(i), static_cast<uint16_t>(i), static_cast<uint16_t>(aFirstValue), lValue);
```
Use `unsigned` ratehr than `uint16_t` here, both places. %x is unsigned.

yufengwangca commented on src/app/server/EchoHandler.cpp:
```diff
@@ -40,7 +40,7 @@ chip::Protocols::Echo::EchoServer gEchoServer;
 void HandleEchoRequestReceived(chip::Messaging::ExchangeContext * ec, chip::System::PacketBufferHandle && payload)
 {
-    ChipLogProgress(AppServer, "Echo Request, len=%u ... sending response.\n", payload->DataLength());
+    ChipLogProgress(AppServer, "Echo Request, len=%u ... sending response.\n", static_cast<uint16_t>(payload->DataLength()));
```
Converting to uint16_t could drop the value.

For maximum portability and to ensure correctness, you should use the format specifiers defined in <inttypes.h> for fixed-width integers. For uint32_t, the correct format specifier is PRIu32:

ChipLogProgress(AppServer, "Echo Request, len= %" PRIu32 "... sending response.\n", static_cast<uint16_t>(payload->DataLength()));

pidarped replied:
Will change to PRIu32.

yufengwangca commented on src/inet/TCPEndPointImplLwIP.cpp:
```diff
@@ -351,7 +351,7 @@ CHIP_ERROR TCPEndPointImplLwIP::DriveSendingImpl()
                 VerifyOrDie(!startOfUnsent.buffer.IsNull());
 
-                uint16_t bufDataLen = startOfUnsent.buffer->DataLength();
+                uint16_t bufDataLen = static_cast<uint16_t>(startOfUnsent.buffer->DataLength());
```
assert for overrun

pidarped replied:
The reason I left out the pre-check in LwIP is because LwIP packet buffers are limited to uint16 size limit.

tcarmelveilleux commented on src/system/tests/TestSystemPacketBuffer.cpp:
```diff
@@ -329,8 +329,18 @@ void PacketBufferTest::PrepareTestBuffer(BufferConfiguration * config, int flags
         config->handle->next = nullptr;
     }
     config->handle->payload = config->payload_ptr;
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    if (config->init_len > UINT16_MAX)
+    {
+        printf("Max Length exceeded for LwIP based systems\n");
+        exit(EXIT_FAILURE);
+    }
```
This has to be a VerifyOrDie.

bzbarsky-apple commented on src/lib/core/tests/TestTLV.cpp:
```diff
@@ -277,7 +277,7 @@ void TestBufferContents(nlTestSuite * inSuite, const System::PacketBufferHandle
     while (!buf.IsNull())
     {
-        uint16_t len = buf->DataLength();
+        uint32_t len = static_cast<uint32_t>(buf->DataLength());
```
Can `len` not be `size_t` here?

pidarped replied:
I think I need to change the passed in argument 'expectedLen' to size_t in order to do that because of all the checks and arithmetic involved in lines below. This seemed like a lesser change and safe as well. But I can check to see if I can change that arg to make len as a size_t.

bzbarsky-apple replied:
Followup is fine it it's too much work.

bzbarsky-apple commented on src/platform/Zephyr/BLEManagerImpl.cpp:
```diff
@@ -658,7 +658,7 @@ bool BLEManagerImpl::SendIndication(BLE_CONNECTION_OBJECT conId, const ChipBleUU
     params->func = HandleTXIndicated;
     params->data = pBuf->Start();
-    params->len  = pBuf->DataLength();
+    params->len  = static_cast<uint16_t>(pBuf->DataLength());
```
Should error out if incoming buffer too big?

bzbarsky-apple commented on src/platform/Zephyr/BLEManagerImpl.cpp:
```diff
@@ -837,7 +837,7 @@ ssize_t BLEManagerImpl::HandleC3Read(struct bt_conn * conId, const struct bt_gat
 
     return bt_gatt_attr_read(conId, attr, buf, len, offset, sInstance.c3CharDataBufferHandle->Start(),
-                             sInstance.c3CharDataBufferHandle->DataLength());
+                             static_cast<uint16_t>(sInstance.c3CharDataBufferHandle->DataLength()));
```
Do we know the cast is safe?

bzbarsky-apple commented on src/platform/android/BLEManagerImpl.cpp:
```diff
@@ -384,7 +384,8 @@ bool BLEManagerImpl::SendWriteRequest(BLE_CONNECTION_OBJECT conId, const Ble::Ch
     SuccessOrExit(err);
 
-    err = JniReferences::GetInstance().N2J_ByteArray(env, pBuf->Start(), pBuf->DataLength(), characteristicDataObj);
+    err = JniReferences::GetInstance().N2J_ByteArray(env, pBuf->Start(), static_cast<uint16_t>(pBuf->DataLength()),
```
Do we know this cast is safe?

bzbarsky-apple commented on src/platform/bouffalolab/common/BLEManagerImpl.cpp:
```diff
@@ -669,7 +669,7 @@ bool BLEManagerImpl::SendIndication(BLE_CONNECTION_OBJECT conId, const ChipBleUU
     params->func = HandleTXIndicated;
     params->data = pBuf->Start();
-    params->len  = pBuf->DataLength();
+    params->len  = static_cast<uint16_t>(pBuf->DataLength());
```
Do we know the casts in this file are safe?

All casts should either come with documentation explaining why they are known to be safe or be preceded by checks that ensure they are in fact safe.

bzbarsky-apple replied:
The documentation needs to be _in the code_.

bzbarsky-apple commented on src/platform/mbed/BLEManagerImpl.cpp:
```diff
@@ -1007,7 +1007,7 @@ bool BLEManagerImpl::SendIndication(BLE_CONNECTION_OBJECT conId, const ChipBleUU
                   conId, att_handle, pBuf->DataLength());
 
-    mbed_err = gatt_server.write(att_handle, pBuf->Start(), pBuf->DataLength(), false);
+    mbed_err = gatt_server.write(att_handle, pBuf->Start(), static_cast<uint16_t>(pBuf->DataLength()), false);
```
I'm going to stop commenting on the "is this safe?" bit.  Please fix all of those.

pidarped replied:
For BLE, the max buffer length is UINT16_MAX.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -1575,7 +1575,7 @@ CHIP_ERROR CASESession::HandleSigma3a(System::PacketBufferHandle && msg)
 
     const uint8_t * buf   = msg->Start();
-    const uint16_t bufLen = msg->DataLength();
+    const uint32_t bufLen = static_cast<uint32_t>(msg->DataLength());
```
Can bufLen not be size_t?

pidarped replied:
This one can be size_t without any other perturbations I think. Will change.

bzbarsky-apple commented on src/system/SystemPacketBuffer.cpp:
%u is the wrong format for uint32_t.

bzbarsky-apple commented on src/system/SystemPacketBuffer.cpp:
```diff
@@ -140,7 +141,7 @@ void PacketBufferHandle::InternalRightSize()
     const uint8_t * const start   = mBuffer->ReserveStart();
     const uint8_t * const payload = mBuffer->Start();
-    const uint16_t usedSize       = static_cast<uint16_t>(payload - start + mBuffer->len);
+    const size_t usedSize         = static_cast<size_t>(static_cast<uint32_t>(payload - start) + mBuffer->len);
```
Why the intermediate cast to uint32_t?  This doesn't make sense.

pidarped replied:
This is because we are adding the difference to mBuffer->len which is a size_t and unsigned. payload - start is guaranteed to be positive, so this cast should be fine.

bzbarsky-apple replied:
None of that answers my question about why the intermediate cast is needed.  What happens if this is written as: `usedSize = (payload - start) + mBuffer->len;` without any casts?

bzbarsky-apple commented on src/system/SystemPacketBuffer.cpp:
```diff
@@ -207,25 +208,36 @@ void PacketBuffer::SetStart(uint8_t * aNewStart)
     ptrdiff_t lDelta = aNewStart - static_cast<uint8_t *>(this->payload);
-    if (lDelta > this->len)
-        lDelta = this->len;
+    if (lDelta > static_cast<int32_t>(this->len))
+        lDelta = static_cast<int32_t>(this->len);
```
What I would expect is that we would check for negative lDelta, perhaps, but then cast lDelta to size_t to do the compare, instead of casting our length.

bzbarsky-apple commented on src/system/SystemPacketBuffer.cpp:
```diff
@@ -207,25 +208,36 @@ void PacketBuffer::SetStart(uint8_t * aNewStart)
-    this->len     = static_cast<uint16_t>(static_cast<ptrdiff_t>(this->len) - lDelta);
-    this->tot_len = static_cast<uint16_t>(static_cast<ptrdiff_t>(this->tot_len) - lDelta);
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    this->len     = static_cast<uint16_t>(static_cast<int32_t>(this->len) - lDelta);
+    this->tot_len = static_cast<uint16_t>(static_cast<int32_t>(this->tot_len) - lDelta);
```
Why are we no longer casting to ptrdiff_t here?

pidarped replied:
ptrdiff_t is not easily inter-castable with size_t on some platforms, IIRC. Hence the change.

bzbarsky-apple replied:
Then shouldn't we cast to ssize_t, instead of throwing in 32-bit assumptions?

bzbarsky-apple commented on src/system/SystemPacketBuffer.cpp:
```diff
@@ -207,25 +208,36 @@ void PacketBuffer::SetStart(uint8_t * aNewStart)
         aNewLen = kMaxDataLen;
 
-    ptrdiff_t lDelta = static_cast<ptrdiff_t>(aNewLen) - static_cast<ptrdiff_t>(this->len);
+    int32_t lDelta = static_cast<int32_t>(aNewLen) - static_cast<int32_t>(this->len);
```
Why is this not using ptrdiff_t?  That seems like the right type to use.

pidarped replied:
ptrdiff_t is not easily inter-castable with size_t on some platforms, IIRC. Hence the change.

bzbarsky-apple replied:
This is why ssize_t exists, no?

bzbarsky-apple commented on src/system/SystemPacketBuffer.cpp:
```diff
@@ -207,25 +208,36 @@ void PacketBuffer::SetStart(uint8_t * aNewStart)
+    int32_t lDelta = static_cast<int32_t>(aNewLen) - static_cast<int32_t>(this->len);
 
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    VerifyOrDieWithMsg(aNewLen < UINT16_MAX, chipSystemLayer, "LwIP buffer length cannot exceed UINT16_MAX");
```
<, or <=?

pidarped replied:
Will change to <=.

bzbarsky-apple commented on src/system/TLVPacketBufferBackingStore.cpp:
```diff
@@ -31,7 +31,7 @@ namespace System {
 {
     bufStart = mHeadBuffer->Start();
-    bufLen   = mHeadBuffer->DataLength();
+    bufLen   = static_cast<uint32_t>(mHeadBuffer->DataLength());
```
Needs to be checking the cast is safe, erroring if not?

This is all going to come back and bite us sometime.  Please fix all such callsites.

bzbarsky-apple commented on src/transport/SecureMessageCodec.cpp:
```diff
@@ -43,22 +43,22 @@ CHIP_ERROR Encrypt(const CryptoContext & context, CryptoContext::ConstNonceView
 
     uint8_t * data    = msgBuf->Start();
-    uint16_t totalLen = msgBuf->TotalLength();
+    uint32_t totalLen = static_cast<uint32_t>(msgBuf->TotalLength());
```
We should be checking this is safe!

bzbarsky-apple commented on src/transport/SecureMessageCodec.cpp:
```diff
@@ -43,22 +43,22 @@ CHIP_ERROR Encrypt(const CryptoContext & context, CryptoContext::ConstNonceView
 
-    VerifyOrReturnError(CanCastTo<uint16_t>(totalLen + taglen), CHIP_ERROR_INTERNAL);
-    msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen));
+    VerifyOrReturnError(CanCastTo<size_t>(totalLen + taglen), CHIP_ERROR_INTERNAL);
```
This is no longer doing correct checks.  Back when totalLen and taglen were both uint16_t, we knew we could add without overflow, since it would add as (4-byte) int.

But now that we have 32-bit things here, we have to be a lot more careful with our additions.  Please fix this.

bzbarsky-apple commented on src/transport/SecureMessageCodec.cpp:
```diff
@@ -69,7 +69,7 @@ CHIP_ERROR Decrypt(const CryptoContext & context, CryptoContext::ConstNonceView
 
     uint8_t * data = msg->Start();
-    uint16_t len   = msg->DataLength();
+    uint32_t len   = static_cast<uint32_t>(msg->DataLength());
```
Why is the cast safe?

bzbarsky-apple commented on src/transport/raw/MessageHeader.cpp:
```diff
@@ -364,9 +364,9 @@ CHIP_ERROR PacketHeader::EncodeBeforeData(const System::PacketBufferHandle & buf
+CHIP_ERROR PayloadHeader::Encode(uint8_t * data, size_t size, uint16_t * encode_size) const
 {
-    VerifyOrReturnError(size >= EncodeSizeBytes(), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(size >= static_cast<size_t>(EncodeSizeBytes()), CHIP_ERROR_INVALID_ARGUMENT);
```
Why is the cast needed?  What type does EncodeSizeBytes() return?

bzbarsky-apple commented on src/inet/TCPEndPointImplSockets.cpp:
```diff
@@ -483,7 +483,7 @@ CHIP_ERROR TCPEndPointImplSockets::DriveSendingImpl()
     while (!mSendQueue.IsNull())
     {
-        uint16_t bufLen = mSendQueue->DataLength();
+        uint32_t bufLen = static_cast<uint32_t>(mSendQueue->DataLength());
```
What was the error?  There shouldn't be an error in that case...

bzbarsky-apple commented on src/inet/TCPEndPointImplSockets.cpp:
```diff
@@ -496,14 +496,14 @@ CHIP_ERROR TCPEndPointImplSockets::DriveSendingImpl()
         }
 
-        if (lenSentRaw < 0 || lenSentRaw > bufLen)
+        if (lenSentRaw < 0 || bufLen < static_cast<uint32_t>(lenSentRaw))
```
I don't see how casting to size_t can possibly complain but casting to uint32_t succeed.  But I'd love to learn more.

bzbarsky-apple commented on src/inet/UDPEndPointImplOpenThread.cpp:
```diff
@@ -237,7 +237,7 @@ CHIP_ERROR UDPEndPointImplOT::SendMsgImpl(const IPPacketInfo * aPktInfo, System:
     VerifyOrExit(message != NULL, error = OT_ERROR_NO_BUFS);
 
-    error = otMessageAppend(message, msg->Start(), msg->DataLength());
+    error = otMessageAppend(message, msg->Start(), static_cast<uint16_t>(msg->DataLength()));
```
My point is this: we are going to throw away the high bytes.  If those are nonzero, we should error out, not silently use a different length.

bzbarsky-apple commented on src/messaging/tests/echo/echo_requester.cpp:
```diff
@@ -184,7 +184,8 @@ void HandleEchoResponseReceived(chip::Messaging::ExchangeContext * ec, chip::Sys
     printf("Echo Response: %" PRIu64 "/%" PRIu64 "(%.2f%%) len=%u time=%.3fs\n", gEchoRespCount, gEchoCount,
-           static_cast<double>(gEchoRespCount) * 100 / static_cast<double>(gEchoCount), payload->DataLength(),
+           static_cast<double>(gEchoRespCount) * 100 / static_cast<double>(gEchoCount),
+           static_cast<uint32_t>(payload->DataLength()),
```
My point is that %u is the wrong format for uint32_t.

bzbarsky-apple commented on src/messaging/tests/echo/echo_responder.cpp:
```diff
@@ -49,7 +49,7 @@ chip::SessionHolder gSession;
 void HandleEchoRequestReceived(chip::Messaging::ExchangeContext * ec, chip::System::PacketBufferHandle && payload)
 {
-    printf("Echo Request, len=%u ... sending response.\n", payload->DataLength());
+    printf("Echo Request, len=%u ... sending response.\n", static_cast<uint32_t>(payload->DataLength()));
```
No, you misunderstood the question.

bzbarsky-apple commented on src/platform/ESP32/nimble/BLEManagerImpl.cpp:
```diff
@@ -575,7 +575,7 @@ bool BLEManagerImpl::SendIndication(BLE_CONNECTION_OBJECT conId, const ChipBleUU
     ESP_LOGD(TAG, "Sending indication for CHIPoBLE TX characteristic (con %u, len %u)", conId, data->DataLength());
 
-    om = ble_hs_mbuf_from_flat(data->Start(), data->DataLength());
+    om = ble_hs_mbuf_from_flat(data->Start(), static_cast<uint16_t>(data->DataLength()));
```
Then that exact reasoning needs to be documented in the code.

Again, all casts either needs checks or clear documentation explaining why they are safe and always will be.
